### PR TITLE
Vector store

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by Sourcery

Replace Redis and Pinecone with Supabase for vector storage and memory management, implementing a new memory system that uses local embeddings with Transformers.js instead of OpenAI embeddings

Enhancements:
- Simplify memory management by consolidating storage to Supabase and removing multi-vendor support
- Switch from OpenAI embeddings to local embeddings using Transformers.js with the all-MiniLM-L6-v2 model
- Implement food memory tracking to maintain context of previous food interactions

Build:
- Add Supabase configuration and database initialization scripts
- Configure Next.js to support external packages for ONNX runtime

Documentation:
- Add local setup instructions for Supabase in README

Chores:
- Remove unused Redis-specific companion key functionality
- Clean up CSS comments in index.css

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `reviews_for_yhkli`.
Run ID: `reviews_for_yhkli/benchmark_2024-10-29T11-27-46_v1-23-0-113-g9fada7a5a-dirty`.

This pull request was cloned from `https://github.com/ykhli/AI-tamago/pull/7`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: benchmark
    enable_security_comments: benchmark
    enable_tests_comments: benchmark
    enable_comment_suggestions: benchmark
    enable_functionality_review: benchmark

    enable_pull_request_summary: benchmark
    enable_review_guide: benchmark

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: benchmark
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/ykhli/AI-tamago/pull/25
- https://github.com/ykhli/AI-tamago/pull/22
- https://github.com/ykhli/AI-tamago/pull/21
- https://github.com/ykhli/AI-tamago/pull/19
- https://github.com/ykhli/AI-tamago/pull/18
- https://github.com/ykhli/AI-tamago/pull/17
- https://github.com/ykhli/AI-tamago/pull/16
- https://github.com/ykhli/AI-tamago/pull/15
- https://github.com/ykhli/AI-tamago/pull/14
- https://github.com/ykhli/AI-tamago/pull/13
- https://github.com/ykhli/AI-tamago/pull/12
- https://github.com/ykhli/AI-tamago/pull/11
- https://github.com/ykhli/AI-tamago/pull/10
- https://github.com/ykhli/AI-tamago/pull/9
- https://github.com/ykhli/AI-tamago/pull/8
- https://github.com/ykhli/AI-tamago/pull/7
- https://github.com/ykhli/AI-tamago/pull/6
- https://github.com/ykhli/AI-tamago/pull/5
- https://github.com/ykhli/AI-tamago/pull/4
- https://github.com/ykhli/AI-tamago/pull/3
- https://github.com/ykhli/AI-tamago/pull/2
- https://github.com/ykhli/AI-tamago/pull/1

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4792
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/414
- dataset_pull_request: https://github.com/ykhli/AI-tamago/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/AI-tamago/pull/1

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This PR refactors the memory management system to use Supabase as the primary vector store, removing Pinecone support and Redis history. It introduces local embedding generation using Transformers.js and adds new interaction tracking functionality. The changes also include database setup and migration files for Supabase.

#### Sequence diagram for feeding interaction with vector search

```mermaid
sequenceDiagram
    participant C as Client
    participant A as API Route
    participant M as MemoryManager
    participant T as Transformers.js
    participant S as Supabase

    C->>A: POST /interact (FEED)
    A->>M: getInstance()
    A->>M: vectorSearch(food)
    M->>T: generateEmbedding(food)
    T-->>M: embedding
    M->>S: match_documents RPC
    S-->>M: matching documents
    M->>S: saveToMemory(comment, metadata)
    M->>S: saveInteraction(FEED, metadata)
    A-->>C: animation and status
```

#### Entity relationship diagram for Supabase tables

```mermaid
erDiagram
    documents ||--o{ llamagotchi_interactions : references
    documents {
        bigint id PK
        text content
        jsonb metadata
        vector embedding
    }
    llamagotchi_interactions {
        varchar interaction
        jsonb metadata
        timestamp ts
    }
    llamagotchi_status {
        int happiness
        int hunger
        int health
        int age
        int other
        timestamp updatedAt
    }
```

#### Class diagram for updated MemoryManager

```mermaid
classDiagram
    class MemoryManager {
        -dbClient: SupabaseClient
        -instance: MemoryManager
        +constructor()
        +init()
        +saveToMemory(content: string, metadata: any)
        +generateEmbedding(content: string)
        +vectorSearch(contentToSearch: string)
        +saveInteraction(interaction: INTERACTION, metadata: any)
        +getInstance(): Promise~MemoryManager~
        +writeToHistory(text: string)
        +readLatestHistory()
        +seedChatHistory()
    }
    note for MemoryManager "Simplified memory management
using Supabase only"
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Replaced dual vector store system with Supabase-only implementation | <ul><li>Removed Pinecone client and related configuration</li><li>Implemented local embedding generation using Transformers.js</li><li>Added new vector search functionality using Supabase RPC</li><li>Removed Redis-based chat history management</li></ul> | `src/app/utils/memory.ts` |
| Added Supabase database setup and configuration | <ul><li>Created migration script for initializing tables and vector store</li><li>Added Supabase configuration file</li><li>Created seed data script</li><li>Added local setup instructions in README</li></ul> | `supabase/migrations/20231029050944_init_tables.sql`<br/>`supabase/config.toml`<br/>`supabase/seed.sql`<br/>`README.md` |
| Updated interaction handling in the application | <ul><li>Modified food interaction tracking</li><li>Added interaction persistence to Supabase</li><li>Updated animation timing and status management</li></ul> | `src/app/api/interact/route.ts`<br/>`src/app/utils/interaction.ts`<br/>`src/components/Tamagotchi.tsx` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.
- **Generate a pull request title:** Write `@sourcery-ai` anywhere in the pull
  request title to generate a title at any time.
- **Generate a pull request summary:** Write `@sourcery-ai summary` anywhere in
  the pull request body to generate a PR summary at any time. You can also use
  this command to specify where the summary should be inserted.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/app/api/interact/route.ts [hunk 1]
```diff
@@ -68,30 +74,61 @@ export async function POST(req: Request) {
...
+        await memoryManager.saveInteraction(
+          INTERACTION.FEED,
+          resultJsonMetadata
+        );
       }
   }
   return NextResponse.json({ animation: JSON.stringify(animation), status });
 }
+
+function updateRecentFood(recent: any, food: string) {
```
Comment (SourceryAI) [inline src/app/api/interact/route.ts:129]: **issue (bug_risk):** The array manipulation logic is incorrect and could lead to bugs

The line `recent = recent.shift()` assigns the removed element back to the array variable. Instead, use `recent.slice(-5).concat(food)` or `recent.push(food); if (recent.length > 5) recent.shift();`

File: src/app/utils/memory.ts [hunk 2]
```diff
@@ -4,147 +4,92 @@ import { PineconeClient } from "@pinecone-database/pinecone";
...
-        .catch((err) => {
-          console.log("WARNING: failed to get vector search results.", err);
-        });
-      return similarDocs;
-    }
-  }
+  public async saveToMemory(content: string, metadata: any) {
+    const embedding = await this.generateEmbedding(content);
+    const supabaseClient = <SupabaseClient>this.dbClient;
+    const error = await supabaseClient
```
Comment (SourceryAI) [inline src/app/utils/memory.ts:33]: **issue:** Missing error handling in saveToMemory

The error response from Supabase should be properly handled. Consider throwing an error or returning a result object that indicates success/failure.

File: README.md [hunk 3]
```diff
@@ -7,3 +8,12 @@
...
 
 All client side tamagotchi code is in Tamagotchi.tsx
+
+## Local Setup
+
+`brew install supabase/tap/supabase`
+
+`supabase start`
+
+To run migrations or reset database -- seed.sql and migrations will run
```
Comment (SourceryAI) [inline README.md:18]: **suggestion (documentation):** Clarify the migrations description

Consider rephrasing to: 'To reset the database and run migrations (including seed.sql):'

```suggestion
To reset the database and run migrations (including seed.sql):
```

File: src/app/api/interact/route.ts [hunk 4]
```diff
@@ -24,18 +29,19 @@ let stats = {
...
   const { interactionType } = await req.json();
   console.debug("interactionType", interactionType);
   let animation = idle;
 
   const model = getModel();
   model.verbose = true;
+  const memoryManager = await MemoryManager.getInstance();
 
   switch (interactionType) {
     case INTERACTION.FEED:
```
Comment (SourceryAI) [inline src/app/api/interact/route.ts:47]: **issue (complexity):** Consider restructuring the FEED handler into smaller, focused functions with clear responsibilities

The FEED handler can be simplified by extracting core operations into focused functions while fixing the recent food tracking. Suggested refactoring:

```typescript
async function generateFood(model: OpenAI, recentFood: string[]) {
  const eatPrompt = PromptTemplate.fromTemplate(`...`); // existing prompt template
  const eatChain = new LLMChain({ llm: model, prompt: eatPrompt });
  const result = await eatChain.call({ stats: JSON.stringify(stats) });
  return JSON.parse(result.text);
}

async function handleMemoryOperations(memoryManager: MemoryManager, food: string, resultJson: any) {
  const foodMemory = await memoryManager.vectorSearch(food);
  await memoryManager.saveToMemory(resultJson.comment, resultJson);
  await memoryManager.saveInteraction(INTERACTION.FEED, resultJson);
}

function updateRecentFood(recent: string[], food: string, maxItems = 5) {
  if (recent.length >= maxItems) {
    recent.shift(); // Remove oldest item
  }
  recent.push(food);
}

// Main handler becomes clearer:
case INTERACTION.FEED: {
  if (stats.eat === FULL) {
    return handleFullState();
  }

  const foodResult = await generateFood(model, recentFood);
  updateRecentFood(recentFood, foodResult.food);
  await handleMemoryOperations(memoryManager, foodResult.food, foodResult);

  return handleFeedingResponse(foodResult);
}
```

This restructuring:
1. Separates concerns into focused functions
2. Fixes the recent food tracking logic
3. Makes the main flow easier to follow
4. Keeps all functionality intact while reducing complexity

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and found some issues that need to be addressed.

**Blocking issues**:
- The array manipulation logic is incorrect and could lead to bugs ([link](https://github.com/sourcery-ai-experiments/AI-tamago/pull/17/files#diff-9a83ccde0992882959c213e4e8f6e041f4f8b50620508dc4ef72b4e3a39dd831R129))

**Overall Comments**:
- The updateRecentFood function has a bug - assigning to 'recent' instead of using array methods properly. Should be recent.shift() rather than recent = recent.shift()

<details>
<summary>Here's what I looked at during the review</summary>

- 🔴 **General issues**: 1 blocking issue, 1 other issue
- 🟢 **Security**: all looks good
- 🟢 **Testing**: all looks good
- 🟡 **Complexity**: 1 issue found
- 🟡 **Documentation**: 1 issue found
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/8cc563f9-dde0-4f01-90e0-ea4401084acc?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Pull request closed without merging**